### ISSUE TITLE
Paramiko RSA key seemingly being mistakenly detected as a DSA key

### ISSUE THREAD

**Issue by nullstream** [ktbyers/netmiko#2827]
Versions in use:
`~/switch-backup$ pip freeze | grep miko
netmiko==4.1.0
paramiko==2.11.0`

Code snippet:
`~/switch-backup$ grep key backups_switches.py
        device_info['use_keys'] = True
        device_info['key_file'] = "~/.ssh/switches"`

When trying to to connect (note SSH manually works just fine):
`
Unknown exception: q must be exactly 160, 224, or 256 bits long
Traceback (most recent call last):
  File "/home/seancody/.local/lib/python3.7/site-packages/paramiko/transport.py", line 2164, in run
    handler(self.auth_handler, m)
  File "/home/seancody/.local/lib/python3.7/site-packages/paramiko/auth_handler.py", line 395, in _parse_service_accept
    sig = self.private_key.sign_ssh_data(blob, algorithm)
  File "/home/seancody/.local/lib/python3.7/site-packages/paramiko/dsskey.py", line 117, in sign_ssh_data
    ).private_key(backend=default_backend())
  File "/home/seancody/.local/lib/python3.7/site-packages/cryptography/hazmat/primitives/asymmetric/dsa.py", line 242, in private_key
    return ossl.load_dsa_private_numbers(self)
  File "/home/seancody/.local/lib/python3.7/site-packages/cryptography/hazmat/backends/openssl/backend.py", line 822, in load_dsa_private_numbers
    dsa._check_dsa_private_numbers(numbers)
  File "/home/seancody/.local/lib/python3.7/site-packages/cryptography/hazmat/primitives/asymmetric/dsa.py", line 283, in _check_dsa_private_numbers
    _check_dsa_parameters(parameters)
  File "/home/seancody/.local/lib/python3.7/site-packages/cryptography/hazmat/primitives/asymmetric/dsa.py", line 275, in _check_dsa_parameters
    raise ValueError("q must be exactly 160, 224, or 256 bits long")
ValueError: q must be exactly 160, 224, or 256 bits long

Traceback (most recent call last):
  File "backup_switches.py", line 49, in <module>
    net_connect = ConnectHandler(**device)
  File "/home/seancody/.local/lib/python3.7/site-packages/netmiko/ssh_dispatcher.py", line 344, in ConnectHandler
    return ConnectionClass(*args, **kwargs)
  File "/home/seancody/.local/lib/python3.7/site-packages/netmiko/base_connection.py", line 434, in __init__
    self._open()
  File "/home/seancody/.local/lib/python3.7/site-packages/netmiko/base_connection.py", line 439, in _open
    self.establish_connection()
  File "/home/seancody/.local/lib/python3.7/site-packages/netmiko/base_connection.py", line 1021, in establish_connection
    self.remote_conn_pre.connect(**ssh_connect_params)
  File "/home/seancody/.local/lib/python3.7/site-packages/paramiko/client.py", line 446, in connect
    passphrase,
  File "/home/seancody/.local/lib/python3.7/site-packages/paramiko/client.py", line 682, in _auth
    self._transport.auth_publickey(username, key)
  File "/home/seancody/.local/lib/python3.7/site-packages/paramiko/transport.py", line 1635, in auth_publickey
    return self.auth_handler.wait_for_response(my_event)
  File "/home/seancody/.local/lib/python3.7/site-packages/paramiko/auth_handler.py", line 245, in wait_for_response
    raise e
  File "/home/seancody/.local/lib/python3.7/site-packages/paramiko/transport.py", line 2164, in run
    handler(self.auth_handler, m)
  File "/home/seancody/.local/lib/python3.7/site-packages/paramiko/auth_handler.py", line 395, in _parse_service_accept
    sig = self.private_key.sign_ssh_data(blob, algorithm)
  File "/home/seancody/.local/lib/python3.7/site-packages/paramiko/dsskey.py", line 117, in sign_ssh_data
    ).private_key(backend=default_backend())
  File "/home/seancody/.local/lib/python3.7/site-packages/cryptography/hazmat/primitives/asymmetric/dsa.py", line 242, in private_key
    return ossl.load_dsa_private_numbers(self)
  File "/home/seancody/.local/lib/python3.7/site-packages/cryptography/hazmat/backends/openssl/backend.py", line 822, in load_dsa_private_numbers
    dsa._check_dsa_private_numbers(numbers)
  File "/home/seancody/.local/lib/python3.7/site-packages/cryptography/hazmat/primitives/asymmetric/dsa.py", line 283, in _check_dsa_private_numbers
    _check_dsa_parameters(parameters)
  File "/home/seancody/.local/lib/python3.7/site-packages/cryptography/hazmat/primitives/asymmetric/dsa.py", line 275, in _check_dsa_parameters
    raise ValueError("q must be exactly 160, 224, or 256 bits long")
ValueError: q must be exactly 160, 224, or 256 bits long
`

Checking the key type...
`~/switch-backup$ ssh-keygen -l -f ~/.ssh/switches
3072 SHA256:Ojr3XqoBT8M<REDACTED>2XCAhyXfE sean@locus (RSA)`

Note key specified is definitely RSA.
If I use password based auth it works just fine... this is generated only because use_keys and key_file.

I've tried converting the key to a PEM (as noted in a few bug reports) but this newer version prefers OPENSSH formatted keys which is fine.

Any ideas?

**Comment by ktbyers** [comment]
@nullstream Looks like this Paramiko issue here (which hasn't been fixed yet):

https://github.com/paramiko/paramiko/issues/1839

**Comment by ktbyers** [comment]
It looks like you might be able to temporarily fix it by pinning to Paramiko version 2.8.1.

Let me know if that works.

**Comment by alphabet5** [comment]
I've worked around this by passing pkey to the ConnectHandler.

```
key_file = "~/.ssh/id_rsa"
key_file_expanded = os.path.expanduser(key_file)
key = paramiko.rsakey.RSAKey(filename=key_file_expanded)
info = {
    "device_type": "linux",
    "host": host",
    "username": "myname",
    "use_keys": True,
    "pkey": key,
    "key_file": key_file,
}
```
